File: src/main/java/org/jukeboxmc/math/Location.java
package org.jukeboxmc.math;

import org.jukeboxmc.block.Block;
import org.jukeboxmc.world.Biome;
import org.jukeboxmc.world.Dimension;
import org.jukeboxmc.world.World;
import org.jukeboxmc.world.chunk.Chunk;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class Location extends Vector {

    private World world;
    private float yaw;
    private float pitch;

    public Location( World world, float x, float y, float z, float yaw, float pitch, Dimension dimension ) {
        super( x, y, z );
        this.world = world;
        this.yaw = yaw;
        this.pitch = pitch;
        this.dimension = dimension;
    }

    public Location( World world, float x, float y, float z, float yaw, float pitch ) {
        this( world, x, y, z, yaw, pitch, Dimension.OVERWORLD );
    }

    public Location( World world, float x, float y, float z, Dimension dimension ) {
        this( world, x, y, z, 0, 0, dimension );
    }

    public Location( World world, float x, float y, float z ) {
        this( world, x, y, z, 0, 0, Dimension.OVERWORLD );
    }

    public Location( World world, int x, int y, int z, Dimension dimension ) {
        this( world, x, y, z, 0, 0, dimension );
    }

    public Location( World world, int x, int y, int z ) {
        this( world, x, y, z, 0, 0, Dimension.OVERWORLD );
    }

    public Location( World world, Vector vector, float yaw, float pitch, Dimension dimension ) {
        this( world, vector.getX(), vector.getY(), vector.getZ(), yaw, pitch, dimension );
    }

    public Location( World world, Vector vector, float yaw, float pitch ) {
        this( world, vector.getX(), vector.getY(), vector.getZ(), yaw, pitch, Dimension.OVERWORLD );
    }

    public Location( World world, Vector vector ) {
        this( world, vector.getX(), vector.getY(), vector.getZ(), 0, 0, Dimension.OVERWORLD );
    }

    @Override
    public Location add( float x, float y, float z ) {
        return new Location( this.world, this.x + x, this.y + y, this.z + z, this.yaw, this.pitch, this.dimension );
    }

    @Override
    public Location subtract( float x, float y, float z ) {
        return new Location( this.world, this.x - x, this.y - y, this.z - z, this.yaw, this.pitch, this.dimension );
    }

    @Override
    public Location multiply( float x, float y, float z ) {
        return new Location( this.world, this.x * x, this.y * y, this.z * z, this.yaw, this.pitch, this.dimension );
    }

    @Override
    public Location divide( float x, float y, float z ) {
        return new Location( this.world, this.x / x, this.y / y, this.z / z, this.yaw, this.pitch, this.dimension );
    }

    public World getWorld() {
        return this.world;
    }

    public void setWorld( World world ) {
        this.world = world;
    }

    public float getYaw() {
        return this.yaw;
    }

    public void setYaw( float yaw ) {
        this.yaw = yaw;
    }

    public float getPitch() {
        return this.pitch;
    }

    public void setPitch( float pitch ) {
        this.pitch = pitch;
    }

    public Block getBlock() {
        return this.world.getBlock( this );
    }

    public Biome getBiome() {
        return this.world.getBiome( this.getBlockX(), this.getBlockY(), this.getBlockZ() );
    }

    public Block getBlock( int layer ) {
        return this.world.getBlock( this, layer );
    }

    public Chunk getChunk() {
        return this.world.getChunk( this.getBlockX() >> 4, this.getBlockZ() >> 4, this.dimension );
    }

    public Vector getDirection() {
        double pitch = ( ( this.pitch + 90 ) * Math.PI ) / 180;
        double yaw = ( ( this.yaw + 90 ) * Math.PI ) / 180;
        double x = Math.sin( pitch ) * Math.cos( yaw );
        double z = Math.sin( pitch ) * Math.sin( yaw );
        double y = Math.cos( pitch );
        return new Vector( (float) x, (float) y, (float) z ).normalize();
    }


    @Override
    public String toString() {
        return "Location{" +
                "world=" + this.world.getName() +
                ", x=" + this.x +
                ", y=" + this.y +
                ", z=" + this.z +
                ", yaw=" + this.yaw +
                ", pitch=" + this.pitch +
                ", dimension=" + this.dimension.name() +
                '}';
    }
}

File: JukeboxMC-API/src/main/kotlin/org/jukeboxmc/api/math/Location.kt
package org.jukeboxmc.api.math

import org.jukeboxmc.api.block.Block
import org.jukeboxmc.api.world.Biome
import org.jukeboxmc.api.world.Dimension
import org.jukeboxmc.api.world.World
import org.jukeboxmc.api.world.chunk.Chunk
import kotlin.math.cos
import kotlin.math.sin


class Location : Vector, Cloneable {

    private var world: World
    private var yaw: Float
    private var pitch: Float

    constructor(world: World, x: Float, y: Float, z: Float, yaw: Float, pitch: Float, dimension: Dimension) : super(x, y, z, dimension) {
        this.world = world
        this.yaw = yaw
        this.pitch = pitch
    }

    constructor(world: World, x: Float, y: Float, z: Float, dimension: Dimension) : super(x, y, z, dimension){
        this.world = world
        this.yaw = 0F
        this.pitch = 0F
    }

    constructor(world: World, x: Float, y: Float, z: Float) : super(x, y, z, Dimension.OVERWORLD){
        this.world = world
        this.yaw = 0F
        this.pitch = 0F
    }

    constructor(world: World, x: Int, y: Int, z: Int, yaw: Float, pitch: Float, dimension: Dimension) : super(x, y, z, dimension){
        this.world = world
        this.yaw = yaw
        this.pitch = pitch
    }

    constructor(world: World, x: Int, y: Int, z: Int, dimension: Dimension) : super(x, y, z, dimension){
        this.world = world
        this.yaw = 0F
        this.pitch = 0F
    }

    constructor(world: World, x: Int, y: Int, z: Int) : super(x, y, z, Dimension.OVERWORLD){
        this.world = world
        this.yaw = 0F
        this.pitch = 0F
    }

    constructor(world: World, vector: Vector, yaw: Float, pitch: Float, dimension: Dimension) : super(vector.getBlockX(), vector.getBlockY(), vector.getBlockZ(), dimension){
        this.world = world
        this.yaw = yaw
        this.pitch = pitch
    }

    constructor(world: World, vector: Vector, dimension: Dimension) : super(vector.getBlockX(), vector.getBlockY(), vector.getBlockZ(), dimension){
        this.world = world
        this.yaw = 0F
        this.pitch = 0F
    }

    constructor(world: World, vector: Vector) : super(vector.getBlockX(), vector.getBlockY(), vector.getBlockZ(), Dimension.OVERWORLD){
        this.world = world
        this.yaw = 0F
        this.pitch = 0F
    }

    override fun add(x: Float, y: Float, z: Float): Location {
        return Location(world, this.getX() + x, this.getY() + y, this.getZ() + z, yaw, pitch, this.getDimension())
    }

    override fun subtract(x: Float, y: Float, z: Float): Location {
        return Location(world, this.getX() - x, this.getY() - y, this.getZ() - z, yaw, pitch, this.getDimension())
    }

    override fun multiply(x: Float, y: Float, z: Float): Location {
        return Location(world, this.getX() * x, this.getY() * y, this.getZ() * z, yaw, pitch, this.getDimension())
    }

    override fun divide(x: Float, y: Float, z: Float): Location {
        return Location(world, this.getX() / x, this.getY() / y, this.getZ() / z, yaw, pitch, this.getDimension())
    }

    fun getWorld(): World {
        return world
    }

    fun getYaw(): Float {
        return this.yaw
    }

    fun setYaw(yaw: Float) {
        this.yaw = yaw
    }

    fun getPitch(): Float {
        return this.pitch
    }

    fun setPitch(pitch: Float) {
        this.pitch = pitch
    }

    fun getBlock(): Block {
        return this.world.getBlock(this)
    }

    fun getBlock(layer: Int): Block {
        return this.world.getBlock(this, layer)
    }

    fun getBiome(): Biome {
        return world.getBiome(this, this.getDimension())
    }

     fun getChunk(): Chunk? {
        return world.getChunk(getBlockX() shr 4, getBlockZ() shr 4, this.getDimension())
    }

    fun getLoadedChunk(): Chunk? {
        return world.getLoadedChunk(getBlockX() shr 4, getBlockZ() shr 4, this.getDimension())
    }

    fun getDirection(): Vector {
        val pitch = (pitch + 90) * Math.PI / 180
        val yaw = (yaw + 90) * Math.PI / 180
        val x = sin(pitch) * cos(yaw)
        val z = sin(pitch) * sin(yaw)
        val y = cos(pitch)
        return Vector(x.toFloat(), y.toFloat(), z.toFloat()).normalize()
    }

    override fun clone(): Location {
        val clone: Vector = super<Vector>.clone()
        return Location(world, clone.getX(), clone.getY(), clone.getZ())
    }

    override fun toString(): String {
        return "Location{" +
                "world=" + world.getWorldData().levelName +
                ", x=" + this.getX() +
                ", y=" + this.getY() +
                ", z=" + this.getZ() +
                ", yaw=" + yaw +
                ", pitch=" + pitch +
                ", dimension=" + this.getDimension().name +
                '}'
    }
}

